### PR TITLE
Allow FROUND fix on x64

### PR DESCRIPTION
**Pull request by kklobe**
# Description

I tested the Quake screen size slider on Windows 11 x64 using the Normal core, and this fix is definitely required there, so remove the conditional compilation.

# Manual testing

Tested with Quake on x64 Normal and DynX86 cores, as well as dynrec core on macOS with Apple Silicon.

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Review by johnnovak — APPROVED** [review]
Ship it 😎

**Comment by johnnovak** [comment]
@kklobe actually, could you re-execute the Intel FPU test suite on x64 and post the results before merging?

**Comment by kklobe** [comment]
> @kklobe actually, could you re-execute the Intel FPU test suite on x64 and post the results before merging? Using both the normal and dynamic cores.

CleanShot 2023-12-15 at 16 46 16@2x

**Comment by johnnovak** [comment]
Thanks for that @kklobe, but hang on for a second...

My understanding is that we've achieved 100% conformance (zero failue) in the Intel FPU test earlier this year:
https://github.com/dosbox-staging/dosbox-staging/pull/3216

Are we saying we have regressed with the recent FPU fixes? Can you please help me while I finalise the release notes and get to the bottom of this? We're mentioning in the notes that we've achieved 100% confirmance in the FPU tests... which seems to be no longer the case? 😐

**Comment by Grounded0** [comment]
Our release notes state that we have 0 errors on both cores on x86-64 so i would like to keep it that way.

**Comment by Grounded0** [comment]
I still get 0 errors on both dynamic and normal cores on a 2 week old `main` build.

**Comment by johnnovak** [comment]
> I still get 0 errors on both dynamic and normal cores on a 2 week old `main` CI build.

On ARM?

**Comment by Grounded0** [comment]
On x86-64 of course. ARM always gives us thousands of errors due to lack of extended precision.

**Comment by johnnovak** [comment]
> On x86-64 of course. ARM always gives us thousands of errors due to lack of extended precision.

Yeah ok. @kklobe tested this on his Mac now, check the title bars on his screenshots 😄 

All in all, 80-bit precision conformance is a moot point for us, as we've previously discussed. And options like  -ffast-math in GCC exist for a reason; in many cases (I'd say the majority of the time), you simply don't care about strict 100% IEEE float conformance.

We'll still mention the conformance in the notes if that's true for x86/x64, of course.

**Comment by Grounded0** [comment]
Time to upgrade your ARM boxes:

Core normal:

Screenshot 2023-12-16 at 1 43 02

Core dynamic:

Screenshot 2023-12-16 at 1 44 58

**Comment by Grounded0** [comment]
• Intel Floating Point Conformance Test passes with 0 failures on the x86 hosts

The Intel Floating Point Conformance Test for IEEE 754-1985 now passes with 0 failures on hosts with the x86 family of CPUs on all the cores. The 80-bit extended precision format is truncated to the 64-bit double precision format on non-x86 hosts for performance reasons, and these hosts shouldn't be used for any engineering or financial calculations.

**Comment by johnnovak** [comment]
> Time to upgrade your ARM boxes:

What's that, an M3? I'm confused again because @kklobe just bought a new MacBook which should also be an M3?

**Comment by kklobe** [comment]
> Are we saying we have regressed with the recent FPU fixes? Can you please help me while I finalise the release notes and get to the bottom of this? We're mentioning in the notes that we've achieved 100% confirmance in the FPU tests... which seems to be no longer the case? 😐

The FPU rounding fix for the Quake screen slider #3161 added 8 more failures to the rounding section.

**Comment by Grounded0** [comment]
I'm not sure if a joke or not but let me post my system details:

2017 15" MacBook Pro with AMD Radeon Pro 560

**Comment by kklobe** [comment]
Normal core on x64 Windows 11 has the same failures as my M3 screenshots above. Dynamic core passes all of them.

**Comment by Grounded0** [comment]
> Normal core on x64 Windows 11 has the same failures as my M3 screenshots above. Dynamic core passes all of them.

Oh dear.

**Comment by kklobe** [comment]
> > Normal core on x64 Windows 11 has the same failures as my M3 screenshots above. Dynamic core passes all of them.
> 
> Oh dear.

Which is expected, since MSVC cannot compile inline assembly on x64, so the normal core ends up using `fpu_instructions.h`, just like non-Intel builds.

**Comment by Grounded0** [comment]
Linux also gets 0 and 0 too so it would be important to find out what's the story on Windows. Very nasty words omitted.

**Comment by johnnovak** [comment]
Yeah, as I said, I don't care much about those performance tests, to be honest. Fixing problems in actual software matters. Pretty sure for 100 people annoyed not being able to set 100% view size in Quake there a single guy staring at the output of MCPDIAG 😄 So while the results will be interesting (perhaps), if this fixes real-world problems, in it goes.

**Comment by Grounded0** [comment]
Sorry i messed up something.

**Comment by Grounded0** [comment]
Let's just stop trying to set the house on fire with days to go on a release m'kay. 😁

**Review by kcgen — APPROVED** [review]
:+1:

**Comment by Grounded0** [comment]
> Existing x86/x64-non-MSVC-builds should not be impacted by this at all.

Okay merge away then if we bypass this.

**Comment by johnnovak** [comment]
> The default Windows install uses MSYS2 build which gets us 0 and 0.
> 
> Sorry about editing your post by accident kklobe. Exactly the type of stuff i warned would happen when these guys reinstated my maintainer rights here.
> 
> Pinging @kcgen

No problem. We'll just revoke it if you can't behave 😆 It happens even to the best of us to make mistakes, all good.

**Comment by kklobe** [comment]
> The default Windows install uses MSYS2 build which gets us 0 and 0.

Yes, the MSYS2 toolchain (and pretty much any non-MSVC x64 compiler) can do inline assembly, so the x87 instructions get compiled in for the normal core properly.

The dynamic x86 (`core_dyn_x86`) core can just emit the proper extended precision opcodes, so dynamic core on any x64 build should work fine. The dynamic non-x86 (`core_dynrec`) core uses the 64-bit precision code in `fpu_instructions.h`.

So, to summarize:

- All x64 builds should pass on dynamic core.
- All non-MSVC x64 builds should pass on normal core.
- Everything else should get the M3 results I posted above.

**Comment by johnnovak** [comment]
> > The default Windows install uses MSYS2 build which gets us 0 and 0.
> 
> Yes, the MSYS2 toolchain (and pretty much any non-MSVC x64 compiler) can do inline assembly, so the x87 instructions get compiled in for the normal core properly.
> 
> The dynamic x86 (`core_dyn_x86`) core can just emit the proper extended precision opcodes, so dynamic core on any x64 build should work fine. The dynamic non-x86 (`core_dynrec`) core uses the 64-bit precision code in `fpu_instructions.h`.
> 
> So, to summarize:
> 
>     1. All x64 builds should pass on dynamic core.
> 
>     2. All non-MSVC x64 builds should pass on normal core.
> 
>     3. Everything else should get the M3 results I posted above.
> 
> 
> None of these FPU fixes should affect 1 or 2 (file a bug if otherwise).

Thanks, that's what I need @kklobe -- focused, succinct, precise info 🎉

**Comment by kcgen** [comment]
@kklobe , I think we're OK to merge, right?

---

It's too bad Intel's conformance tests were construed as Intel-specific bit-exact results, as opposed to comparing IEEE thresholds. 

2.0 + 1.0 = 2.99999999987 : Intel ✔️
2.0 + 1.0 = 3.00000000000 : non-Intel ❌ 

I'd much rather see magnitude of error, and see things like 1x10^-13 instead of green and red boxes. 

No doubt, mcpdiag was the perfect software to dissuade system vendors like Compaq from buying non-Intel chips. _"You don't want your customers getting wrong FPU results, now do you?"_

**Comment by kklobe** [comment]
@kcgen yep I don't have anything else to add.

**Comment by kklobe** [comment]
> @kklobe , I think we're OK to merge, right?
> 
> It's too bad Intel's conformance tests were construed as Intel-specific bit-exact results, as opposed to comparing IEEE thresholds.
> 
> 2.0 + 1.0 = 2.99999999987 : Intel ✔️ 2.0 + 1.0 = 3.00000000000 : non-Intel ❌
> 

It's even worse than that: note the Sine only has 8 passes. Here's are the "failures":

```log
2023-12-16 14:32:13.999 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:13.999 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:13.999 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:13.999 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.000 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(0.000000000000000) = 0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.001 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.002 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.002 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.002 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.002 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.002 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
2023-12-16 14:32:14.002 | FPU: FSIN(-0.000000000000000) = -0.000000000000000
```

So I guess "it's zero, but it's not INTEL zero!"
